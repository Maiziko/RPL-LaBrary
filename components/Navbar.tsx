import Link from 'next/link';

const Navbar = () => {
  return (
    <div className='h-[0px]'>
      <nav className="fixed top-0 left-0 right-0 bg-primary p-3">
        <div>
          <div className='flex'>
            <button className="cursor-pointer mr-[40px] ml-[50px]" id="burger-menu">
              <img src="/icon/sandwichMenu.png" alt="Burger Menu" />
            </button>
            <div className='mr-[70px]'>
              <Link href="/koleksibuku">
                  <img src="/images/libraryLogo.svg" alt="Library Logo" />
              </Link>
            </div>
            <div className="w-[600px] h-10 pl-4 flex items-center justify-center rounded-full px-4 duration-300 cursor-pointer bg-white">
              <img src="/images/searchBar.svg" alt="Search Bar" />
              <input className="text-[15px] ml-4 w-full bg-transparent focus:outline-none" placeholder="Cari di LaBrary..." />
            </div>
            <div className="mr-[50px] pl-[207px]">
              <img src="/images/ProfilePictureSideMenu.svg" alt="Profile Picture" />
            </div>
          </div>
        </div>
      </nav>
    </div>
  );
};

export default Navbar;
