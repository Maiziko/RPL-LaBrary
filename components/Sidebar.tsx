// components/Sidebar.tsx
import { useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import Link from 'next/link';
import { supabase } from '../supabase';

interface ProfileData {
  full_name: string;
  email: string;
  avatar_url: string;
}

// Fungsi untuk menampilkan sidebar
function showSidebar() {
  const sidebar = document.getElementById("sidebar") as HTMLElement | null;

  if (sidebar) {
    sidebar.classList.remove("transform", "translate-x-[-240px]");
  }
}
// Fungsi untuk menyembunyikan sidebar
function hideSidebar() {
  const sidebar = document.getElementById("sidebar") as HTMLElement | null;
  if (sidebar) {
    sidebar.classList.add("transform", "translate-x-[-240px]");
  }
}


const Sidebar = () => {
  const router = useRouter();
  const [profileData, setProfileData] = useState<ProfileData | null>(null);

  const fetchProfileData = async () => {
    try {
      // Ambil ID pengguna dari penyimpanan lokal (localStorage)
      const userId = localStorage.getItem('userId');

      if (userId) {
        const { data, error } = await supabase
          .from('profiles')
          .select('full_name, email, avatar_url')
          .eq('id', userId)
          .single();

        if (error) {
          console.error('Error fetching profile data:', error.message);
        } else {
          setProfileData(data);
        }
      }
    } catch (error) {
      console.error('Error fetching profile data:', (error as Error).message);
    }
  };

  const handleLogout = async () => {
    try {
      const { error }: { error: Error | null } = await supabase.auth.signOut();
      if (error) {
        console.error('Error logging out:', error.message);
      } else {
        // Redirect to the home page after successful logout
        router.push('/');
        console.log('User logged out successfully');
      }
    } catch (error) {
      console.error('Error logging out:', (error as Error).message);
    }
  };

  // Fungsi untuk menutup sidebar
  const hideSidebar = () => {
    const sidebar = document.getElementById("sidebar") as HTMLElement | null;
    if (sidebar) {
      sidebar.classList.add("transform", "translate-x-[-240px]");
    }
  };

  const handleOutsideClick = (event: MouseEvent) => {
    const sidebar = document.getElementById("sidebar") as HTMLElement | null;
    const target = event.target as HTMLElement;

    // Periksa apakah yang diklik adalah di luar sidebar
    if (sidebar && !sidebar.contains(target)) {
      hideSidebar();
    }
  };
  
  useEffect(() => {
    const burgerMenuButton = document.getElementById("burger-menu");
    if (burgerMenuButton) {
      burgerMenuButton.addEventListener("click", showSidebar);
    }

    const closeSidebarButton = document.getElementById("close-sidebar");
    if (closeSidebarButton) {
      closeSidebarButton.addEventListener("click", hideSidebar);
    }

    document.addEventListener("mousedown", handleOutsideClick);

    return () => {
      if (burgerMenuButton) {
        burgerMenuButton.removeEventListener("click", showSidebar);
      }
      if (closeSidebarButton) {
        closeSidebarButton.removeEventListener("click", hideSidebar);
      }
      document.removeEventListener("mousedown", handleOutsideClick);
    };
  }, []);

  useEffect(() => {
    fetchProfileData();
  }, []);


  return (
    <div className="fixed rounded-r-[5px] top-0 left-0 lg:left-0 lg:w-[240px] w-60 h-screen bg-primary text-gray-100 transform translate-x-[-240px] transition-transform duration-300" id="sidebar">
      {/* Tombol Tutup Sidebar */}
      <button className="bg-gray-green py-4 px-3 rounded-full cursor-pointer absolute top-5 left-4" id="close-sidebar">
        <img src="icon/BackButtonSideBar.png" className='w-[30px]' alt="" />
      </button>
      
      {/* Informasi Profil */}
      <div className="mt-[70px] text-center">
        {profileData && (
          <>
           <img
            src={profileData.avatar_url || '/images/ProfilePictureSideMenu.svg'}
            className='w-16 h-16 mx-auto rounded-full'
            alt=""
          />
            <h2 className="text-xl font-semibold mt-2">{profileData.full_name}</h2>
            <p className="text-sm mt-1">{profileData.email}</p>
          </>
        )}
      </div>

      {/* Menu Sidebar */}
      <ul className="mt-14">
        <li className="">
          <Link href="/koleksibuku" className="flex px-10 py-3 w-[240px] text-white group hover:bg-[#C5D0D2] rounded-r-full hover:text-strong-orange text-14 transition-all duration-300">
            <div className="w-[24px] h-[24px] relative group">
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-0 absolute inset-0">
                <img src="/images/Homeoff.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-100">
                <img src="/images/Homeon.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
            </div>
            <span className="pl-5 my-auto text-sm group-hover:text-[#C86F43]">Home</span>
          </Link>
        </li>
        <li className="">
          <Link href="/listpeminjaman" className="flex px-10 py-3 w-[240px] text-white group hover:bg-[#C5D0D2] rounded-r-full hover:text-strong-orange text-14 transition-all duration-300">
            <div className="w-[24px] h-[24px] relative group">
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-0 absolute inset-0">
                <img src="/images/listPeminjamanOff.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-100">
                <img src="/images/ListPeminjamanOn.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
            </div>
            <span className="pl-5 my-auto text-sm group-hover:text-[#C86F43]">List Peminjaman</span>
          </Link>
        </li>
        <li className="">
          <Link href="/koleksibuku" className="flex px-10 py-3 w-[240px] text-white group hover:bg-[#C5D0D2] rounded-r-full hover:text-strong-orange text-14 transition-all duration-300">
            <div className="w-[24px] h-[24px] relative group">
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-0 absolute inset-0">
                <img src="/images/koleksiBukuOffsvg.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-100">
                <img src="/images/koleksiBukuOn.png" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
            </div>
            <span className="pl-5 my-auto text-sm group-hover:text-[#C86F43]">Koleksi Buku</span>
          </Link>
        </li>
        <li className="">
          <Link href="/riwayatbuku" className="flex px-10 py-3 w-[240px] text-white group hover:bg-[#C5D0D2] rounded-r-full hover:text-strong-orange text-14 transition-all duration-300">
            <div className="w-[24px] h-[24px] relative group">
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-0 absolute inset-0">
                <img src="/images/riwayatOff.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
              <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-100">
                <img src="/images/riwayatOn.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
              </div>
            </div>
            <span className="pl-5 my-auto text-sm group-hover:text-[#C86F43]">Riwayat</span>
          </Link>
        </li>
      </ul>

      <div className='mt-[200px]'>
        <button
          className="flex px-10 py-3 w-[240px] text-white group hover:bg-[#C5D0D2] rounded-r-full hover:text-strong-orange text-14 transition-all duration-300"
          onClick={handleLogout}
        >
          <div className="w-[24px] h-[24px] relative group">
            <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-0 absolute inset-0">
              <img src="/icon/LogOutICON.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
            </div>
            <div className="bg-cover bg-center w-[24px] h-[24px] transition-opacity group-hover:opacity-100">
              <img src="/icon/LogOutHoverIcon.svg" alt="" className="group-hover:filter group-hover:grayscale-0" />
            </div>
          </div>
          <span className="pl-5 my-auto text-sm group-hover:text-[#C86F43]">Log Out</span>
        </button>
      </div>
    </div>
  );
};
export default Sidebar;