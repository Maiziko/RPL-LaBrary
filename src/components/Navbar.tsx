import Link from 'next/link';
import { useState, useEffect } from 'react';
import { supabase } from '../../supabase';

const Navbar = () => {
  const [avatarUrl, setAvatarUrl] = useState(null);

  useEffect(() => {
    // Ambil ID pengguna dari penyimpanan lokal (localStorage)
    const userId = localStorage.getItem('userId');

    if (userId) {
      // Ambil data profil dari Supabase berdasarkan ID pengguna
      const fetchProfile = async () => {
        const { data, error } = await supabase
          .from('profiles') // Ganti dengan nama tabel yang sesuai
          .select('avatar_url') // Ganti dengan nama kolom yang sesuai
          .eq('id', userId) // Filter berdasarkan ID pengguna
          .single();

        if (data) {
          setAvatarUrl(data.avatar_url);
        }
      };

      fetchProfile();
    }
  }, []); 

  return (
    <div className='h-[0px]'>
      <nav className="fixed top-0 left-0 right-0 bg-primary p-3">
        <div>
          <div className='flex'>
            <button className="cursor-pointer mr-[40px] ml-[50px]" id="burger-menu">
              <img src="/icon/sandwichMenu.png" alt="Burger Menu" />
            </button>
            <div className='mr-[70px]'>
              <Link href="/koleksibuku">
                  <img src="/images/libraryLogo.svg" alt="Library Logo" />
              </Link>
            </div>
            <div className="w-[600px] h-10 pl-4 flex items-center justify-center rounded-full px-4 duration-300 cursor-pointer bg-white">
              <img src="/images/searchBar.svg" alt="Search Bar" />
              <input className="text-[15px] ml-4 w-full bg-transparent focus:outline-none" placeholder="Cari di LaBrary..." />
            </div>
            <div className="mr-[50px] pl-[207px]">
              {avatarUrl ? (
                <img src={avatarUrl} className='w-[40px]' alt="Profile Picture" />
              ) : (
                <img src="/images/ProfilePictureSideMenu.svg" alt="Default Profile Picture" />
              )}
            </div>
          </div>
        </div>
      </nav>
    </div>
  );
};

export default Navbar;
