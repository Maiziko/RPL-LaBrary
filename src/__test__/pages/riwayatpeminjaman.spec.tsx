import RiwayatPeminjaman from "../../../pages/riwayatpeminjaman";
import Navbar from "@src/components/Navbar";
import Sidebar from "@src/components/Sidebar";
import { render } from "@testing-library/react";
import { SetStateAction } from "react";

describe ("riwayatbuku page", () => {
    it("render koleksibuku Page", () => {
        const page = render(<RiwayatPeminjaman/>);
        expect(page).toMatchSnapshot();
    });
    it("render Navbar", () => {
        const page = render(<Navbar setSearchValue={function (value: SetStateAction<string>): void {
            throw new Error("Function not implemented.");
        } }/>);
        expect(page).toMatchSnapshot();
    });
    it("render Sidebar Page", () => {
        const page = render(<Sidebar/>);
        expect(page).toMatchSnapshot();
    });
});