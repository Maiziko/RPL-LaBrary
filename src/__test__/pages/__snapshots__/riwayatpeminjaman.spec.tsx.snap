// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`riwayatbuku page render Navbar 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-[0px]"
      >
        <nav
          class="fixed top-0 left-0 right-0 bg-primary p-3"
        >
          <div>
            <div
              class="flex"
            >
              <button
                class="cursor-pointer mr-[40px] ml-[50px]"
                id="burger-menu"
              >
                <img
                  alt="Burger Menu"
                  src="/icon/sandwichMenu.png"
                />
              </button>
              <div
                class="mr-[70px]"
              >
                <a
                  href="/koleksibuku"
                >
                  <img
                    alt="Library Logo"
                    src="/images/libraryLogo.svg"
                  />
                </a>
              </div>
              <div
                class="w-[600px] h-10 pl-4 flex items-center justify-center rounded-full px-4 duration-300 cursor-pointer bg-white"
              >
                <img
                  alt="Search Bar"
                  src="/images/searchBar.svg"
                />
                <input
                  class="text-[15px] ml-4 w-full bg-transparent focus:outline-none"
                  placeholder="Cari di LaBrary..."
                  value=""
                />
              </div>
              <div
                class="mr-[50px] pl-[207px]"
              >
                <img
                  alt="Default Profile Picture"
                  src="/images/ProfilePictureSideMenu.svg"
                />
              </div>
            </div>
          </div>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-[0px]"
    >
      <nav
        class="fixed top-0 left-0 right-0 bg-primary p-3"
      >
        <div>
          <div
            class="flex"
          >
            <button
              class="cursor-pointer mr-[40px] ml-[50px]"
              id="burger-menu"
            >
              <img
                alt="Burger Menu"
                src="/icon/sandwichMenu.png"
              />
            </button>
            <div
              class="mr-[70px]"
            >
              <a
                href="/koleksibuku"
              >
                <img
                  alt="Library Logo"
                  src="/images/libraryLogo.svg"
                />
              </a>
            </div>
            <div
              class="w-[600px] h-10 pl-4 flex items-center justify-center rounded-full px-4 duration-300 cursor-pointer bg-white"
            >
              <img
                alt="Search Bar"
                src="/images/searchBar.svg"
              />
              <input
                class="text-[15px] ml-4 w-full bg-transparent focus:outline-none"
                placeholder="Cari di LaBrary..."
                value=""
              />
            </div>
            <div
              class="mr-[50px] pl-[207px]"
            >
              <img
                alt="Default Profile Picture"
                src="/images/ProfilePictureSideMenu.svg"
              />
            </div>
          </div>
        </div>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
